tsdate/prior.py

```diff
@@ -462,6 +462,14 @@ def __init__(self, tree_sequence, *, progress=False, allow_unary=False):
             lambda: defaultdict(lambda: defaultdict(base.FLOAT_DTYPE))
         )
 
+        if not allow_unary:
+            for tree in self.ts.trees():
+                # final element corresponding to the tree’s virtual_root - don't consider.
+                if any(tree.num_children_array[:-1] == 1):
```

Comment:
This LGTM: but do you know if it takes up much time for large tree sequences (I wonder if `np.any` is faster?)?

If it's only a tiny fraction of the time of the total (which I suspect), then it's fine.

Reply:
I've just tried it on https://zenodo.org/record/5512994/files/hgdp_tgp_sgdp_high_cov_ancients_chr2_q.dated.trees.tsz. It takes 4 seconds to iterate through the trees, 3 mins 40 to check unaryness using `np.any`, and > 1 hour using `any()`. I guess that's because it's looking through *all* the nodes (including those not in the tree) for *every* tree. That could be billions of operations.

Reply:
I think the following incremental-type approach seems to be the fastest: it only check those nodes whose children change (i.e. appear as a parent in an edge) for each tree. This takes just under one and a half minutes on my laptop for a 138MB tree sequence, which I think is acceptable? Do you want to check it does the right thing, @barneyhill ?

```python
for tree, ediff in zip(self.ts.trees(), self.ts.edge_diffs()):
    changed = {e.parent for edges in (ediff.edges_out, ediff.edges_in) for e in edges}
    if (tree.num_children_array[list(changed)] == 1).any():
        raise ValueError
```

Reply:
Opened as a discussion in https://github.com/tskit-dev/tskit/discussions/2865 - note that we want to call these "locally unary" nodes.

Reply:
Hi @barneyhill - do you want to change this to the `has_locally_unary_nodes()` method in https://github.com/tskit-dev/tskit/discussions/2865, then I can merge this? I want to get a tsdate release out next week.